### PR TITLE
Fault tolerant Reaper

### PR DESCRIPTION
**Pull request by adejanovski**
Allows Reaper instances to run concurrently on the Cassandra storage backend.
Leader election with LWTs controls which instance can repair a segment, and metrics (pending compactions, active repairs, etc...) can be stored/retrieved from the db to handle cases where an instance cannot reach the whole cluster through JMX.

Adds a "local mode" when JMX is only accessible from localhost. This case requires to have one instance of Reaper running on each Cassandra node.

### CODE REVIEW

**Comment by michaelsembwever** [comment]
This PR has "driven through the hedge"¹

We need the branch to be redone from scratch, with no master commits merge-duplicated in it.
Let me know if you'd like help.

–– 
¹ https://www.youtube.com/watch?v=AJ-CpGsCpM0

File: src/main/java/com/spotify/reaper/ReaperApplication.java
Comment (tlpsonarqube) [inline src/main/java/com/spotify/reaper/ReaperApplication.java:78]: MINOR Make this IP "127.0.0.1" address configurable. [rule](https://sonarcloud.io/coding_rules#rule_key=squid%3AS1313)

File: src/main/java/com/spotify/reaper/cassandra/JmxConnectionFactory.java
Comment (tlpsonarqube) [inline src/main/java/com/spotify/reaper/cassandra/JmxConnectionFactory.java]: MINOR Remove this unused import 'com.spotify.reaper.ReaperApplication'. [rule](https://sonarcloud.io/coding_rules#rule_key=squid%3AUselessImportCheck)

File: src/main/java/com/spotify/reaper/cassandra/JmxConnectionFactory.java
Comment (tlpsonarqube) [inline src/main/java/com/spotify/reaper/cassandra/JmxConnectionFactory.java]: MINOR Make this IP "127.0.0.1" address configurable. [rule](https://sonarcloud.io/coding_rules#rule_key=squid%3AS1313)
MINOR Introduce a new variable instead of reusing the parameter "host". [rule](https://sonarcloud.io/coding_rules#rule_key=squid%3AS1226)

File: src/main/java/com/spotify/reaper/cassandra/JmxProxy.java
Comment (tlpsonarqube) [inline src/main/java/com/spotify/reaper/cassandra/JmxProxy.java]: MINOR Remove this unused import 'com.spotify.reaper.ReaperApplication'. [rule](https://sonarcloud.io/coding_rules#rule_key=squid%3AUselessImportCheck)

File: src/main/java/com/spotify/reaper/core/RepairSchedule.java
Comment (tlpsonarqube) [inline src/main/java/com/spotify/reaper/core/RepairSchedule.java]: MINOR Remove this duplicated import. [rule](https://sonarcloud.io/coding_rules#rule_key=squid%3AUselessImportCheck)

File: src/main/java/com/spotify/reaper/service/RepairRunner.java
Comment (tlpsonarqube) [inline src/main/java/com/spotify/reaper/service/RepairRunner.java:37]: MINOR Remove this unused import 'com.spotify.reaper.ReaperApplication'. [rule](https://sonarcloud.io/coding_rules#rule_key=squid%3AUselessImportCheck)

File: src/main/java/com/spotify/reaper/service/SegmentRunner.java
Comment (tlpsonarqube) [inline src/main/java/com/spotify/reaper/service/SegmentRunner.java]: MAJOR Either re-interrupt this method or rethrow the "InterruptedException". [rule](https://sonarcloud.io/coding_rules#rule_key=squid%3AS2142)

File: src/main/java/com/spotify/reaper/service/SegmentRunner.java
Comment (tlpsonarqube) [inline src/main/java/com/spotify/reaper/service/SegmentRunner.java]: MAJOR Remove this useless assignment to local variable "hostMetrics". [rule](https://sonarcloud.io/coding_rules#rule_key=squid%3AS1854)

File: src/main/java/com/spotify/reaper/service/SegmentRunner.java
Comment (tlpsonarqube) [inline src/main/java/com/spotify/reaper/service/SegmentRunner.java]: MAJOR Extract this nested try block into a separate method. [rule](https://sonarcloud.io/coding_rules#rule_key=squid%3AS1141)

File: src/main/java/com/spotify/reaper/storage/CassandraStorage.java
Comment (tlpsonarqube) [inline src/main/java/com/spotify/reaper/storage/CassandraStorage.java]: MAJOR Remove this "return" statement or make it conditional. [rule](https://sonarcloud.io/coding_rules#rule_key=squid%3AS1751)

File: src/main/java/com/spotify/reaper/storage/CassandraStorage.java
Comment (tlpsonarqube) [inline src/main/java/com/spotify/reaper/storage/CassandraStorage.java]: MAJOR Merge this if statement with the enclosing one. [rule](https://sonarcloud.io/coding_rules#rule_key=squid%3AS1066)

**Comment by tlpsonarqube** [comment]
SonarQube analysis reported 12 issues
* ![MAJOR][MAJOR] 5 major
* ![MINOR][MINOR] 7 minor

Watch the comments in this conversation to review them.

[MAJOR]: https://sonarsource.github.io/sonar-github/severity-major.png 'Severity: MAJOR'
[MINOR]: https://sonarsource.github.io/sonar-github/severity-minor.png 'Severity: MINOR'

File: src/main/java/com/spotify/reaper/cassandra/JmxProxy.java
Comment (tlpsonarqube) [inline src/main/java/com/spotify/reaper/cassandra/JmxProxy.java]: MINOR Remove this unused import 'com.spotify.reaper.ReaperApplication'. [rule](https://sonarcloud.io/coding_rules#rule_key=squid%3AUselessImportCheck)

**Comment by tlpsonarqube** [comment]
SonarQube analysis reported 11 issues
* ![MAJOR][MAJOR] 5 major
* ![MINOR][MINOR] 6 minor

Watch the comments in this conversation to review them.

[MAJOR]: https://sonarsource.github.io/sonar-github/severity-major.png 'Severity: MAJOR'
[MINOR]: https://sonarsource.github.io/sonar-github/severity-minor.png 'Severity: MINOR'

File: src/main/java/com/spotify/reaper/resources/ClusterResource.java
Comment (tlpsonarqube) [inline src/main/java/com/spotify/reaper/resources/ClusterResource.java]: MINOR Remove this unused import 'java.util.stream.IntStream'. [rule](https://sonarcloud.io/coding_rules#rule_key=squid%3AUselessImportCheck)

File: src/main/java/com/spotify/reaper/resources/ClusterResource.java
Comment (tlpsonarqube) [inline src/main/java/com/spotify/reaper/resources/ClusterResource.java]: INFO Complete the task associated to this TODO comment. [rule](https://sonarcloud.io/coding_rules#rule_key=squid%3AS1135)

File: src/main/java/com/spotify/reaper/resources/ClusterResource.java
Comment (tlpsonarqube) [inline src/main/java/com/spotify/reaper/resources/ClusterResource.java]: MINOR Remove this unused import 'java.util.concurrent.ExecutionException'. [rule](https://sonarcloud.io/coding_rules#rule_key=squid%3AUselessImportCheck)

File: src/main/java/com/spotify/reaper/resources/ClusterResource.java
Comment (tlpsonarqube) [inline src/main/java/com/spotify/reaper/resources/ClusterResource.java]: MINOR Replace this lambda with a method reference. [rule](https://sonarcloud.io/coding_rules#rule_key=squid%3AS1612)

**Comment by tlpsonarqube** [comment]
SonarQube analysis reported 15 issues
* ![MAJOR][MAJOR] 5 major
* ![MINOR][MINOR] 9 minor
* ![INFO][INFO] 1 info

Watch the comments in this conversation to review them.

[INFO]: https://sonarsource.github.io/sonar-github/severity-info.png 'Severity: INFO'
[MAJOR]: https://sonarsource.github.io/sonar-github/severity-major.png 'Severity: MAJOR'
[MINOR]: https://sonarsource.github.io/sonar-github/severity-minor.png 'Severity: MINOR'

File: src/main/java/com/spotify/reaper/service/SegmentRunner.java
Comment (tlpsonarqube) [inline src/main/java/com/spotify/reaper/service/SegmentRunner.java]: MAJOR Change this instance-reference to a static reference. [rule](https://sonarcloud.io/coding_rules#rule_key=squid%3AS2209)

File: src/main/java/com/spotify/reaper/service/SegmentRunner.java
Comment (tlpsonarqube) [inline src/main/java/com/spotify/reaper/service/SegmentRunner.java]: MINOR Remove this unused import 'com.spotify.reaper.resources.view.NodesStatus'. [rule](https://sonarcloud.io/coding_rules#rule_key=squid%3AUselessImportCheck)

File: src/main/java/com/spotify/reaper/service/SegmentRunner.java
Comment (tlpsonarqube) [inline src/main/java/com/spotify/reaper/service/SegmentRunner.java]: MINOR Replace this lambda with a method reference. [rule](https://sonarcloud.io/coding_rules#rule_key=squid%3AS1612)

File: src/main/java/com/spotify/reaper/resources/ClusterResource.java
Comment (tlpsonarqube) [inline src/main/java/com/spotify/reaper/resources/ClusterResource.java]: MINOR Replace this lambda with a method reference. [rule](https://sonarcloud.io/coding_rules#rule_key=squid%3AS1612)

**Comment by michaelsembwever** [comment]
rebased off master

**Comment by michaelsembwever** [comment]
The remaining to do on fault-tolerance…

- [x] • looking at postpone outside of leader-election (it can happen!), 
- [x] • making segment reads+writes use local_quorum (is the client driver configured with the correct load balancing policy? does the leader-election mean segment reads+writes are bound to a DC? and not to confuse target DCs with backend storage DCs…),
- [x] • a circle test that tests multi-DC setup, by using a 2DC 6node ccm.

**Comment by michaelsembwever** [comment]
> looking at postpone outside of leader-election (it can happen!),

solved by https://github.com/thelastpickle/cassandra-reaper/pull/112/commits/7c86c0ea4a811bf18a2c98e6765f7f091b153db8

>  making segment reads+writes use local_quorum…

solved by https://github.com/thelastpickle/cassandra-reaper/pull/112/commits/a35349f90f4be462201c2080c8f8f74fb79062b8

**Review by michaelsembwever — APPROVED** [review]

**Comment by adejanovski** [comment]
Let's not merge it as is. 
Latest commits broke the multi region scenario, and we're working on a proper design here.

**Comment by michaelsembwever** [comment]
> Latest commits broke the multi region scenario…

to be precise: it's commit 5d382e6 "Add methods to JmxProxy so know which datacenters each host is in" that fixed one multi-DC scenario but broke another (which was already broken in the branch…)

**Pull request merged**
